framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1292,82 +1365,95 @@ int pcm_wait(struct pcm *pcm, int timeout)
 		if (msg.msgId == AUDIO_MSG_DEQUEUE) {
 			apb = (struct ap_buffer_s *)msg.u.pPtr;
-			pcm->next_buf = apb;
-			pcm->next_size = apb->nbytes;
-			pcm->buf_ptr--;
-			return 1;
+			apb->flags &= ~AUDIO_APB_OUTPUT_ENQUEUED;
+			apb->curbyte = 0;
+			count++;
+		}
+	}
+
+	if (count > 0) {
+		return 1;
+	}
+
+	/* If there were no buffers in the queue, wait for codec to put a buffer on the queue */
+	if (timeout > 0) {
+		/* Use the timeout given by application */
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		st_time.tv_nsec += timeout * MILLI_TO_NANO;
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+	} else {
+		/* Application did not give a timeout value
+		The behavior here depends on whether mq was opened with O_NONBLOCK
+		This might block until a msg becomes available on the queue */
+		size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
+	}
+
+	if (size != sizeof(msg)) {
+		if (errno == ETIMEDOUT) {
+			oops(pcm, errno, "TIMEOUT while watiting for deque message from kernel");
+			return 0;
 		} else {
-			return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
+			return oops(pcm, errno, "Interrupted while waiting for deque message from kernel");
 		}
 	}
+	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
+		/* If you find a dequeue message in the queue,
+		update the apb flags and curbyte and increment the count*/
+		apb = (struct ap_buffer_s *)msg.u.pPtr;
+		apb->flags &= ~AUDIO_APB_OUTPUT_ENQUEUED;
+		apb->curbyte = 0;
+		return 1;
+	}
+
+	return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
 }
 
 int pcm_mmap_transfer(struct pcm *pcm, const void *buffer, unsigned int bytes)
 {
 	int err = 0, frames, avail;
 	unsigned int offset = 0, count;
 
+	/* No data to be transferred */
 	if (bytes == 0) {
 		return 0;
 	}
 
-	count = pcm_bytes_to_frames(pcm, bytes);
+	/* Start the PCM, if required */
+	if (pcm->flags & PCM_IN && !pcm->running) {
+		if (pcm_start(pcm) < 0) {
+			return oops(pcm, errno, "start error\n");
+		}
+	}
 
+	count = pcm_bytes_to_frames(pcm, bytes);
 	while (count > 0) {
-
 		/* get the available space for writing new frames */
 		avail = pcm_avail_update(pcm);
 		if (avail < 0) {
 			return oops(pcm, ENOMEM, "cannot determine available mmap frames");
 		}
 
-		/* start the audio if we reach the threshold */
-		if (!pcm->running) {
-			if (pcm_start(pcm) < 0) {
-				return oops(pcm, errno, "start error: avail 0x%x\n", avail);
-			}
-		}
-
 		/* sleep until we have space to write new frames */
-		if (pcm->running && (unsigned int)avail == 0) {
-			int time = -1;
-
-			/* We will not support NOIRQ flag presently */
-#if 0
-			if (pcm->flags & PCM_NOIRQ)
-				time = (pcm->buffer_size - avail - pcm->mmap_control->avail_min)
-					   / pcm->noirq_frames_per_msec;
-#endif
-			err = pcm_wait(pcm, time);
+		if ((unsigned int)avail == 0) {
```

Comment:
In the proposed change, 0 is also returned by pcm_avail_update when pcm object is null(line 1333), so I think we should change that to a negative value.

Reply:
Fixed.